[ZOOKEEPER-1177] Add the memory optimized watch manager for concentrate watches scenario

lvfangmin opened the pull request:
The current HashSet based WatcherManager will consume more than 40GB memory when 
creating 300M watches.

This patch optimized the memory and time complexity for concentrate watches scenario, compared to WatchManager, both the memory consumption and time complexity improved a lot. I'll post more data later with micro benchmark result.

Changed made compared to WatchManager:
* Only keep path to watches map
* Use BitSet to save the memory used to store watches
* Use ConcurrentHashMap and ReadWriteLock instead of synchronized to reduce lock retention
* Lazily clean up the closed watchers

lvfangmin commented:
@hanm the code complained by Findbug are all correct and expected:

1. System.exit if create watch manager failed in DataTree
It's using the factory to create the watch manager based on the class name, in case we cannot initialize the class due to invalid classname, we have to exit.

2. BitHashSet.elementCount not synchronized in iterator()
It cannot guarantee thread safe during iteration inside this function, so it's marked as non thread safe in the comment, and the caller needs to and is synchronizing it during iterating.

3. Synchronize on AtomicInteger
In the code, we synchronize on it to do wait/notify, but not relying on the synchronization to control the  AtomicInteger value update, so it's used correctly.

anmolnar commented:
@lvfangmin In which case you need to add exceptions to `config/findbugsExcludeFile.xml`.

maoling commented:
@lvfangmin  interesting.where can we find a benchmark which shows how this pr improves the memory?

lvfangmin commented:
Based on internal benchmark, this may save more than 95% memory usage on concentrate watches scenario. I'm working on adding some micro benchmark.

In the original Jira, @phunt also added some basic test, which could show you some ideas about how much memory it's going to save.

The current HashMap based watch manager uses lots of memory due to the overhead of storing each entry (32 Bytes for each entry), the object reference and the duplicated path strings.

Use BitSet without revert index could reduce those overhead and make it more memory efficient.

lvfangmin commented:
There is a PR in the Jira few years ago, which uses BitMap as well, but it sacrificed the performance on triggering watches, this patch improved that and uses lazy clean up for cleaning those watchers who has closed the session.

anmolnar commented:
This is going to be a huge improvement for Accumulo for example that is a heavy user of watchers. I'm going to allocate some capacity to review these new patches.

lvfangmin commented:
Added JMH micro benchmark for the watch manager:

* It shows **big win** for the watch heavy cases, with the current implementation, it uses more than 50MB memory to store 1M watches, with WatchManagerOptimized it only uses around 0.2MB. 
* It also makes add and trigger watches more efficient, since WatchManagerOptimized doesn't maintain the reverse map. 
* In sparse watches use case, the WatchManagerOptimized is expected to use a bit more memory because it needs extra effort to maintain those bit set. In the test it shows around 10% more memory usage.

Here are more result about the throughput/latency related with WatchManager: 

```
Benchmark                               (pathCount)    (watchManagerClass)  (watcherCount)  Mode  Cnt   Score    Error  Units
WatchBench.testAddConcentrateWatch            10000           WatchManager             N/A  avgt    9   5.382 ±  0.968  ms/op
WatchBench.testAddConcentrateWatch            10000  WatchManagerOptimized             N/A  avgt    9   0.696 ±  0.133  ms/op
WatchBench.testAddSparseWatch                 10000           WatchManager           10000  avgt    9   4.889 ±  1.585  ms/op
WatchBench.testAddSparseWatch                 10000  WatchManagerOptimized           10000  avgt    9   4.794 ±  1.068  ms/op
WatchBench.testTriggerConcentrateWatch            1           WatchManager               1  avgt    9  ≈ 10⁻⁴           ms/op
WatchBench.testTriggerConcentrateWatch            1           WatchManager            1000  avgt    9   0.037 ±  0.002  ms/op
WatchBench.testTriggerConcentrateWatch            1  WatchManagerOptimized               1  avgt    9  ≈ 10⁻⁴           ms/op
WatchBench.testTriggerConcentrateWatch            1  WatchManagerOptimized            1000  avgt    9   0.025 ±  0.001  ms/op
WatchBench.testTriggerConcentrateWatch         1000           WatchManager               1  avgt    9   0.048 ±  0.003  ms/op
WatchBench.testTriggerConcentrateWatch         1000           WatchManager            1000  avgt    9  71.838 ±  4.043  ms/op
WatchBench.testTriggerConcentrateWatch         1000  WatchManagerOptimized               1  avgt    9   0.079 ±  0.002  ms/op
WatchBench.testTriggerConcentrateWatch         1000  WatchManagerOptimized            1000  avgt    9  26.135 ±  0.223  ms/op
WatchBench.testTriggerSparseWatch             10000           WatchManager           10000  avgt    9   1.207 ±  0.035  ms/op
WatchBench.testTriggerSparseWatch             10000  WatchManagerOptimized           10000  avgt    9   1.321 ±  0.019  ms/op
```

You can try the following command to run the micro benchmark:
```
$ ant clean package
$ ant clean package -buildfile zookeeper-contrib/zookeeper-contrib-fatjar/build.xml
$ java -jar build/contrib/fatjar/zookeeper-dev-fatjar.jar jmh
```

@maoling @anmolnar hope this gives you a more vivid comparison between the old and new watch manager implementation.

anmolnar commented on src/java/main/org/apache/zookeeper/server/watch/WatchManagerOptimized.java:
```diff
@@ -0,0 +1,355 @@
+     * @param deadWatchers the watchers need to be removed
+     */
+    @Override
+    public void processDeadWatchers(Set<Integer> deadWatchers) {
```
Don't you need to acquire read lock from `addRemovePathRWLock` here?
Both Maps are being modified here and I think there's a possibility for race with `addWatch()` method.

anmolnar commented on src/java/main/org/apache/zookeeper/server/watch/WatchManagerOptimized.java:
```diff
@@ -0,0 +1,355 @@
+
+        int triggeredWatches = 0;
+
+        // Avoid race condition between dead watcher cleaner in
```
I'm not sure if this comment is right. I can't see `WatcherCleaner` anywhere synchronize on the path's BitHashSet. Please correct me if I'm wrong.

anmolnar replied:
Sorry, I got it now.
BitHashSet's remove method is synchronized, so the cleaner implicitly gets a lock on that.

anmolnar commented on src/java/main/org/apache/zookeeper/server/watch/WatchManagerFactory.java:
```diff
@@ -0,0 +1,48 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class WatchManagerFactory {
```
A quick javadoc would be awesome here.

anmolnar commented on src/java/main/org/apache/zookeeper/server/watch/DeadWatcherListener.java:
```diff
@@ -0,0 +1,31 @@
+
+import java.util.Set;
+
+public interface DeadWatcherListener {
```
Please add a few words javadoc here.

anmolnar commented on src/java/main/org/apache/zookeeper/server/util/BitMap.java:
```diff
@@ -0,0 +1,121 @@
+import java.util.BitSet;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+public class BitMap<T> {
```
I think a short javadoc similar to BitHashSet's would useful here.

anmolnar commented on src/java/main/org/apache/zookeeper/server/watch/WatchManagerOptimized.java:
```diff
@@ -0,0 +1,355 @@
+                    continue;
+                }
+
+                w.process(e);
```
Not the scope of this PR, but I've found it surprising that `process()` is not guarded by try-catch. A single watcher could prevent others from triggering by throwing an exception?

anmolnar commented on src/java/test/org/apache/zookeeper/server/DumbWatcher.java:
```diff
@@ -0,0 +1,96 @@
+import org.apache.zookeeper.WatchedEvent;
+import org.apache.zookeeper.proto.ReplyHeader;
+
+public class DumbWatcher extends ServerCnxn {
```
Please consider using mockito.

lvfangmin replied:
I agree from unit test case mock object is easier to maintain than stub ones, but I also need this DumbWatcher in the micro benchmark, I'll put this class somewhere in the code, so the micro benchmark and unit test can share it.

lvfangmin commented on build.xml:
```diff
@@ -119,6 +119,7 @@ xmlns:cs="antlib:com.puppycrawl.tools.checkstyle.ant">
     <property name="test.java.classes" value="${test.java.build.dir}/classes"/>
     <property name="test.src.dir" value="${src.dir}/java/test"/>
     <property name="systest.src.dir" value="${src.dir}/java/systest"/>
+    <property name="bench.src.dir" value="${src.dir}/java/bench"/>
```
Will do.

lvfangmin commented on src/java/main/org/apache/zookeeper/server/watch/WatchManagerOptimized.java:
```diff
@@ -0,0 +1,355 @@
+
+    @Override
+    public boolean removeWatcher(String path, Watcher watcher) {
+        addRemovePathRWLock.writeLock().lock();
```
Actually, this is based on the scenario optimization, like cache hit, when user send removeWatcher request, it's more likely the watcher is actually exist and haven't fired yet, so instead of multiple read lock and API call to check exist then switch to write lock, it's actually cheaper to only call write lock once.

anmolnar commented on src/java/main/org/apache/zookeeper/server/watch/WatchManager.java:
```diff
@@ -46,15 +48,26 @@
     }
 
-    synchronized void addWatch(String path, Watcher watcher) {
+    boolean isDeadWatcher(Watcher watcher) {
```
Taking that into account and the jira fix version, this patch will definitely go into 3.5 as well.

anmolnar commented on src/java/main/org/apache/zookeeper/server/watch/WatchManagerOptimized.java:
```diff
@@ -0,0 +1,355 @@
+    @Override
+    public void removeWatcher(Watcher watcher) {
+        Integer watcherBit;
+        addRemovePathRWLock.writeLock().lock();
```
In which case you need to move `addDeadWatcher()` call inside the critical block.

lvfangmin replied:
Missed this comment last time, what we need here is that, as long as we called addDeadWatcher, there will be no watches added related with this dead watcher. Code executed after line 136 means the watcher will be marked as stale, after we release this lock, any on flying addWatcher for this dead watcher will be rejected, so it guarantees when we call addDeadWatcher there will be no race condition between removing and adding watch.

And I need to move addDeadWatcher out of the locking block, since the WatchCleaner might block on it to avoid OOM issue if the cleaner cannot catch up of cleaning the dead watchers.

anmolnar replied:
Where do you mark the watcher as stale inside the critical block?
It only calls a getter on the `BitIdMap`, right?

lvfangmin replied:
In the caller of removeWatcher, which is in NIOServerCnxn.close and NettyServerCnxn.close.

When the cnxn is closed, it will set stale before call removeCnxn on zkServer, which calls this function sequentially, if we grabbed this lock, it means the cnxn has been marked as stale.

We can explicitly setStale here as well, but I don't think that's necessary.

anmolnar commented:
@lvfangmin Docs for the new caching parameter `zookeeper.bitHashCacheSize`?

lvfangmin commented:
Thanks @nkalmar for the suggestion of bench project position, I was following the same directory as the src/java/systest for now, do you think we can move them together later? I'm not against to move it now.

If we want to move now, just to confirm the directory is src/java/org/apache/zookeeper/bench/ without main folder, right?

nkalmar commented:
systest will go to src/test/java , from my side, you can put the bench in org.apache.zookeeper.test.system . Thinking about it, that's a pretty good place.

No need to create main directory or anything, I will move all the files anyway. Just move the files amongst the others in systest. (Hopefully no package level dependency, I didn't check)

But there's a chance you will have to rebase if this this PR cannot be merged before the movement of all the remaining files as the directory refactor's last step. Sorry about that in advance... I'm not going to be the most popular with that PR :(

lvfangmin commented:
Update based on the comments:

1. add findbugs exclusion
2. add comments for the newly added classes
3. add admin doc for the new JVM options introduced in this diff
4. move the bench project to src/test/java/bench
5. move DumbWatcher to src/java/main dir for share between unit test and bench
6. change to use ExitCode

@nkalmar I moved the bench to src/test/java/bench, which seems more reasonable to me, let me know if you think that's not a good position based on your plan.

lvfangmin commented:
Resolve conflict with latest code on master.

lvfangmin commented:
Rebased to resolve conflict. @anmolnar @hanm @maoling @nkalmar please revisit this PR when you have time.

anmolnar commented:
@lvfangmin Thanks. Sorry for the delay. I'd like to check one more thing before accepting. Bear with me please.

anmolnar commented:
@lvfangmin Just to wrap up the difference between this and original 6-year-old patch on Jira: you've added `deadWatchers` collection and lazy `WatcherCleaner` to avoid the performance penalty on `removeWatches()`.

Is that correct?

lvfangmin commented:
@anmolnar Thanks for reviewing, take your time. 

Here are the main differences between this version and the old one on the Jira:

1. use path to watchers map instead of watcher to paths to improve the performance of trigger watches, we need the lazily dead watcher clean up because of this (main change)
2. better and cleaner implementation, for example, added WatchManagerFactory to easily switch between different watch manager implementation
3. some perf improvement by using HashSet and BitSet to find a balance between memory usage and time complexity
4. fix the watcher leaking issue due to adding dead watcher (we can separate this out if we want to)
5. fix the NettyServerCnxn doesn't de-register itself from watcher manager when the cnxn closed (this is actually fixed recently in #612, so it's not necessary to do it here now)
6. added jmh micro benchmark

anmolnar commented:
I cannot commit, because we still have -1 from @hanm . Waiting for him to approve.
Does it make #612 redundant?

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2223/

hanm commented on src/test/java/bench/org/apache/zookeeper/BenchMain.java:
```diff
@@ -0,0 +1,12 @@
+package org.apache.zookeeper;
```
This file is missing apache license header. This triggers a -1 in last jenkins build.

lvfangmin replied:
Will add it.

hanm commented:
@anmolnar I'll sign this off by end of next Monday if no other issues. 
@lvfangmin great work and thanks for your patience!

hanm commented on zookeeper-docs/src/documentation/content/xdocs/zookeeperAdmin.xml:
```diff
@@ -1025,6 +1025,102 @@ server.3=zoo3:2888:3888</programlisting>
+                <ulink url="https://issues.apache.org/jira/browse/ZOOKEEPER-1179">ZOOKEEPER-1179</ulink> The new watcher
+                manager WatchManagerOptimized will clean up the dead watchers lazily, this config is used to decide how
+                many thread is used in the WatcherCleaner. More thread usually means larger clean up throughput. The
+                default value is 2, which is good enough even for heavy and continuous session closing/receating cases.</para>
```
`closing/recreating`

hanm commented on src/java/main/org/apache/zookeeper/server/watch/WatcherCleaner.java:
```diff
@@ -0,0 +1,176 @@
+                try {
+                    if (deadWatchers.size() < watcherCleanThreshold) {
+                        int maxWaitMs = (watcherCleanIntervalInSeconds +
+                            r.nextInt(watcherCleanIntervalInSeconds / 2 + 1)) * 1000;
```
is there a particular reason of choosing this versus, say exponential backoff?

lvfangmin replied:
Clean up dead watches on large watches ensemble is a heavy work, which might affect the performance, so add jitter to make sure we don't do the lazily clean up at the same time on all the servers in the ensemble.

hanm commented on src/java/main/org/apache/zookeeper/server/watch/WatcherCleaner.java:
```diff
@@ -0,0 +1,176 @@
+                deadWatchers.clear();
+                int total = snapshot.size();
+                LOG.info("Processing {} dead watchers", total);
+						    cleaners.schedule(new WorkRequest() {
```
indentation issue on this line...

hanm commented on src/java/main/org/apache/zookeeper/server/watch/WatcherCleaner.java:
```diff
@@ -0,0 +1,176 @@
+            synchronized (this) {
+                // Snapshot of the current dead watchers
+                final Set<Integer> snapshot = new HashSet<Integer>(deadWatchers);
+                deadWatchers.clear();
```
Is there a particular reason to copy the `deadWatchers`, clear it immediately and then do the work, instead of just operate on `deadWatchers` directly and only clear it after the work is done? I assume the motivation was to free `deadWatchers` earlier so we can pipeline the work: adding more dead watchers while the previous pipeline of cleaning was in progress, but it looks like the new dead watchers will block on `totalDeadWatchers`, which will only be reset after previous dead watchers were cleaned up.

lvfangmin replied:
Clean the dead watchers need to go through all the current watches, which is pretty heavy and may take a second if there are millions of watches, that's why we're doing lazily batch clean up here, we don't want to block addDeadWatcher, which is called from the Cnxn.close while we're doing the clean work.

The totalDeadWatchers is used to avoid OOM when the watcher cleaner cannot catch up (we haven't seen this problem even with heavy reconnecting scenario), it is suggested to be set to something like 1000 * watcherCleanThreshold. The watcherCleanThreshold is used to control the batch size when doing the clean up, there is trade off between GC the dead watcher memory and the time complexity of cleaning up, so we cannot set this too large.

lvfangmin commented on src/java/main/org/apache/zookeeper/server/watch/BitHashSet.java:
```diff
@@ -0,0 +1,156 @@
+        if (elementBit == null) {
+            return false;
+        }
+        return elementBits.get(elementBit);
```
BitSet.get is O(1), check cache doesn't may actually more expensive. 

HashSet is used to optimize the iterating, for example, if there is a single element in this BitHashSet, but the bit is very large, without HashSet we need to go through all the words before return that element, which is not efficient.

hanm replied:
would be good to add a comment at the declaration of `cache` variable, stating that its purpose is to optimize iteration.

lvfangmin commented on src/java/main/org/apache/zookeeper/server/watch/BitHashSet.java:
```diff
@@ -0,0 +1,156 @@
+     * iterate through this set.
+     */
+    @Override
+    public Iterator<Integer> iterator() {
```
It's used in the triggerWatcher with for iterator.

lvfangmin commented on src/java/main/org/apache/zookeeper/server/util/BitMap.java:
```diff
@@ -0,0 +1,125 @@
+        Integer bit = getBit(value);
+        if (bit != null) {
+            return bit;
+        }
```
This BitMap is used by WatchManagerOptimized.watcherBitIdMap, which is used to store watcher to bit mapping.

Add might be called a lot if the same client connection is watching on thousands of even millions of nodes, remove only called once when the session is closed, that's why we optimized to check read lock first in add, but use write lock directly in remove.

lvfangmin commented on src/java/main/org/apache/zookeeper/server/watch/BitHashSet.java:
```diff
@@ -0,0 +1,156 @@
+ * limitations under the License.
+ */
+
+package org.apache.zookeeper.server.watch;
```
At the beginning when we added this class, it was bound with Watcher, but not anymore after refactoring, we can move this to server.util, I'll do that.

lvfangmin commented on src/java/main/org/apache/zookeeper/server/watch/BitHashSet.java:
```diff
@@ -0,0 +1,156 @@
+ */
+public class BitHashSet implements Iterable<Integer> {
+
+    static final long serialVersionUID = 6382565447128283568L;
```
Previously, it was using inheritance instead of composition with HashSet, at that time we added this serialVersionUID, didn't remove this after changing to composition, will remove it.

lvfangmin commented:
@anmolnar #612 was fired because there is user reporting that issue, and it's better to solve it earlier than waiting this diff being reviewed and merged. I'll do the rebase to get rid of that change in this patch.

hanm commented on src/java/main/org/apache/zookeeper/server/watch/WatchManagerOptimized.java:
```diff
@@ -0,0 +1,355 @@
+    @Override
+    public boolean addWatch(String path, Watcher watcher) {
+        boolean result = false;
+        addRemovePathRWLock.readLock().lock();
```
Is the purpose of using a read lock here is to optimize for `addWatch` heavy workloads? Would be good to add a comment here about why choose use a read lock instead of write lock.

lvfangmin replied:
Yes, it's used to improve the read throughput, creating new watcher bit and adding it to the BitHashSet has it's own lock to minimize the lock scope.  I'll add some comments here.

hanm commented on src/java/main/org/apache/zookeeper/server/watch/WatchManagerOptimized.java:
```diff
@@ -0,0 +1,355 @@
+                if (watchers == null) {
+                    watchers = new BitHashSet();
+                    BitHashSet existingWatchers = pathWatches.putIfAbsent(path, watchers);
+                    if (existingWatchers != null) {
```
Is this check necessary, because we are using a read lock here so it's possible for another thread to modify the `pathWatches` while we are here?

lvfangmin replied:
That's correct, reading requests are processed concurrently in CommitProcessor worker service, so it's possible multiple thread might add to pathWatches while we're holding read lock, that's why we need this check here.

hanm commented on src/java/main/org/apache/zookeeper/server/watch/WatchManagerOptimized.java:
```diff
@@ -0,0 +1,355 @@
+
+    @Override
+    public boolean containsWatcher(String path, Watcher watcher) {
+        BitHashSet watchers = pathWatches.get(path);
```
Would be good to add a comment here regarding why no synchronization is required here.

hanm commented:
thanks @lvfangmin for detailed reply.
just had another review pass over some files i forgot to look last time. overall looks good. will sign this off once all review comments are addressed. thanks

lvfangmin commented:
Added more comments about the locking where @anmolnar and @hanm asked during review, it will make the future reference easier as well. Also corrected the typo and unused import.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2278/

hanm commented:
got a ["green" build](https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2285/testReport/) (minors the known failed test `testReconfigRemoveClientFromStatic` discussed at [ZOOKEEPER-2847](https://issues.apache.org/jira/browse/ZOOKEEPER-2847)).

hanm commented:
merged to master. great work @lvfangmin !

plsnotracking commented:
Is it worth it to back port this to 3.4? @hanm @lvfangmin

vivekpatani commented:
Is it worth it to back port this to 3.4? @hanm @lvfangmin

vivekpatani commented:
Is it worth it to back port this to 3.4? @hanm @lvfangmin

eolivelli commented:
@vivekpatani sorry but 3.4 is end of life .
Consider moving to 3.6.x.
We ate going to release 3.7.0 so even 3.5.x is becoming old.
